README.md
# rehype-gist

A rehype plugin for embedding static GitHub Gists in your HTML and MD/X content.

![GitHub package.json version](https://img.shields.io/github/package-json/v/goldenpathtechnologies/rehype-gist)
[![Release](https://github.com/goldenpathtechnologies/rehype-gist/actions/workflows/release.yml/badge.svg)](https://github.com/goldenpathtechnologies/rehype-gist/actions/workflows/release.yml)
[![codecov](https://codecov.io/gh/goldenpathtechnologies/rehype-gist/branch/main/graph/badge.svg?token=fcEjr5yu6q)](https://codecov.io/gh/goldenpathtechnologies/rehype-gist)
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg)](https://conventionalcommits.org)
![GitHub](https://img.shields.io/github/license/goldenpathtechnologies/rehype-gist)

## Demo

`gist:darylwright/75332f27a6e9bff70bc0406114570829`

The above will render a static GitHub Gist on the [`rehype-gist` homepage](https://rehype-gist.goldenpath.ca/).
More demos can be found [here](https://rehype-gist.goldenpath.ca/demo/).

## Installation

PNPM:

```bash
pnpm add rehype-gist
```

NPM:

```bash
npm install rehype-gist
```

Yarn:

```bash
yarn install rehype-gist
```

### Styling

`rehype-gist` depends on GitHub's Gist stylesheet. Run the following script to download a copy of this
stylesheet to use in your project:

#### Unix shell

```bash
curl -s https://gist.github.com/darylwright/75332f27a6e9bff70bc0406114570829.json | jq .stylesheet | xargs -n 1 curl -s -o gist-embed.css
```

#### PowerShell

```powershell
iwr $(iwr https://gist.github.com/darylwright/75332f27a6e9bff70bc0406114570829.json | ConvertFrom-Json).stylesheet -OutFile gist-embed.css
```

Alternatively, you can browse any Gist's JSON in your browser and find the URL of the stylesheet.

```json {8}
{
  "description": "This is an example code snippet used for test data",
  "public": true,
  "created_at": "2023-06-01T09:50:34.000-03:00",
  "files": ["gist-test.ts"],
  "owner": "darylwright",
  "div": "...",
  "stylesheet": "https://github.githubassets.com/assets/gist-embed-38d69729aad4.css"
}
```

The default theme of the downloaded stylesheet is GitHub Light. You can change the default appearance
by importing themes from elsewhere. [`gist-syntax-themes`](https://github.com/lonekorean/gist-syntax-themes)
is a good source of attractive themes for your Gists. Ensure all stylesheets are imported globally in
your project.

The Gists as styled with the default stylesheet have a white border that doesn't align well with the
content. Copy and paste the following CSS in your project's global stylesheet to remove this border
for all your embedded Gists.

```css
.gist table,
.gist th,
.gist td {
    border-width: 0 !important;
}
body .gist .gist-file {
    border-width: 0;
}
```

## Configuration

| Name                   | Type               | Default | Description                                                                                                                             |
|------------------------|--------------------|---------|-----------------------------------------------------------------------------------------------------------------------------------------|
| replaceParentParagraph | boolean            | true    | Transform the parent `<p>` of the qualifying `<code>` element into a Gist. Doesn't transform if the Gist `<code>` element has siblings. |
| omitCodeBlocks         | boolean            | true    | Do not transform `<code>` element whose parent is a `<pre>` element.                                                                    |
| classNames             | string or string[] |         | A list of class names to apply to the root element of the rendered Gist.                                                                |

### Example

```typescript
import { unified } from "unified";
import rehypeParse from "rehype-parse";
import rehypeStringify from "rehype-stringify";
import rehypeGist from "rehype-gist";

const output = await unified()
  .use(rehypeParse)
  .use(rehypeGist, {
    replaceParentParagraph: false,
    omitCodeBlocks: false,
    classNames: [`border-2`, `my-4`],
  })
  .use(rehypeStringify)
  .process(`
<!doctype html>
<html lang="en">
  <head>
    <title>Transform inline code to static GitHub Gist</title>
  </head>
  <body>
    <code>gist:darylwright/75332f27a6e9bff70bc0406114570829</code>
  </body>
</html>
  `);
```

## Notes

This plugin should be declared before any others that transform `<code>` tags, or the Gists will not
be rendered.

Highlights and removed lines can only be applied to one file at a time in a Gist. Multi-file Gists
can still be displayed without highlights and removed lines.

## Issues

Please report any issues with this software
[here](https://github.com/goldenpathtechnologies/rehype-gist/issues). If you would like to contribute to
this project, feel free to fork it and send a pull request. Note that this project is governed by a
[code of conduct](https://github.com/goldenpathtechnologies/rehype-gist/blob/main/CODE_OF_CONDUCT.md).

## Acknowledgements

This project was inspired by [`gatsby-remark-embed-gist`](https://github.com/weirdpattern/gatsby-remark-embed-gist)
by [Patricio Treviño](https://www.weirdpattern.com/).

## License

This project is [MIT](https://github.com/goldenpathtechnologies/rehype-gist/blob/main/LICENSE.md)
licensed.


CHANGELOG.md
## 1.0.0 (2023-06-21)


### Features

* initial release ([a13d675](https://github.com/goldenpathtechnologies/rehype-gist/commit/a13d675bf94427597097f6ab2ad98007261d2cd7))


test/util.test.ts
import axios from "axios";
import type { Element, Properties } from "hast";
import { isElement } from "hast-util-is-element";
import { visit } from "unist-util-visit";
import getGistElement, {
  convertFilenameToAttributeFormat,
  GistUri,
  parseGistUri,
  getGistElementFromHtml,
  getGistVisitorPredicate,
  getGistTreeFromHtml,
  getGistVisitor,
  getCodeLineIdPrefix, RehypeGistOptions, convertInlineCodeToGist, isValidGist, elementHasOneValidChild,
} from "../src/util.js";
import { singleFileGist } from "./data/index.js";

describe(`parseGistUri`, () => {
  it(`Returns undefined when the URI does not have Gist protocol`, () => {
    const result = parseGistUri(
      `darylwright/f33e544d005767520266637c82fe5604`,
    );

    expect(result).toBeUndefined();
  });

  it(`Correctly parses the username and gistId`, () => {
    const result = parseGistUri(
      `gist:darylwright/f33e544d005767520266637c82fe5604`,
    );

    expect(result).not.toBeUndefined();
    expect(result.username).toBe(`darylwright`);
    expect(result.gistId).toBe(`f33e544d005767520266637c82fe5604`);
  });

  it(`Returns undefined when the username and gistId is separated incorrectly`, () => {
    const result = parseGistUri(
      `gist:darylwright?f33e544d005767520266637c82fe5604`,
    );

    expect(result).toBeUndefined();
  });

  it(`Correctly parses 'file' query string parameter`, () => {
    const result = parseGistUri(
      `gist:darylwright/f33e544d005767520266637c82fe5604?file=Example.txt`,
    );

    expect(result).not.toBeUndefined();
    expect(result.file).toBe(`Example.txt`);
  });

  it(`Correctly parses 'lines' query string parameter`, () => {
    const result = parseGistUri(
      `gist:darylwright/f33e544d005767520266637c82fe5604?lines=1,3`,
    );

    expect(result).not.toBeUndefined();
    expect(result.lines).toEqual([1, 3]);
  });

  it(`Correctly parses 'highlights' query string parameter`, () => {
    const result = parseGistUri(
      `gist:darylwright/f33e544d005767520266637c82fe5604?highlights=2-4,6`,
    );

    expect(result).not.toBeUndefined();
    expect(result.highlights).toEqual([2, 3, 4, 6]);
  });

  it(`Returns undefined when the gistId and query string parameters are separated incorrectly`, () => {
    const result = parseGistUri(
      `gist:darylwright/f33e544d005767520266637c82fe5604/lines=1,3`,
    );

    expect(result).toBeUndefined();
  });

  it(`Returns undefined when there are excess URI delimeters`, () => {
    const result0 = parseGistUri(
      `gist:darylwright/f33e544d005767520266637c82fe5604/randomtext?lines=1,3`,
    );
    const result1 = parseGistUri(
      `gist:darylwright/f33e544d005767520266637c82fe5604?randomtext?lines=1,3`,
    );

    expect(result0).toBeUndefined();
    expect(result1).toBeUndefined();
  });

  it(`Correctly handles empty query strings`, () => {
    const result = parseGistUri(
      `gist:darylwright/f33e544d005767520266637c82fe5604?`,
    );

    expect(result).not.toBeUndefined();
    expect(result.file).toBeUndefined();
    expect(result.lines).toEqual([]);
    expect(result.highlights).toEqual([]);
  });
});

describe(`convertFilenameToAttributeFormat`, () => {
  it(`Removes leading period from filenames`, () => {
    const result = convertFilenameToAttributeFormat(`.file`);

    expect(result).toBe(`file`);
  });

  it(`Replaces non-alphanumeric and underscore characters with hyphens`, () => {
    const result = convertFilenameToAttributeFormat(`name#of+file09.extension`);

    expect(result).toBe(`name-of-file09-extension`);
  });

  it(`Converts all alphanumeric characters to lowercase`, () => {
    const result = convertFilenameToAttributeFormat(`FILE`);

    expect(result).toBe(`file`);
  });
});

describe(`getGistTreeFromHtml`, () => {
  it(`Disregards sibling nodes that are not elements`, () => {
    const html = `
      <div><div>
      This is plain text.
    `;

    expect(() => getGistTreeFromHtml(html)).not.toThrowError();
  });

  it(`Returns an element from raw HTML`, () => {
    const result = getGistTreeFromHtml(`<div id="test"></div>`);

    expect(isElement(result)).toBeTruthy();
    expect(result.tagName).toBe(`div`);
    expect(result.properties.id).not.toBeUndefined();
    expect(result.properties.id).toBe(`test`);
  });
});

describe(`getGistVisitorPredicate`, () => {
  const defaultTestUri: GistUri = {
    gistId: `justsomerandomwords`,
    username: `username`,
    file: `filename`,
    lines: [],
    highlights: [],
  };
  const defaultPredicateFn = getGistVisitorPredicate(defaultTestUri);

  it(`Returns false when node is not a 'tr' element`, () => {
    const node: Element = { type: `element`, tagName: `table`, children: [] };
    const result = defaultPredicateFn(node);

    expect(result).toBeFalsy();
  });

  it(`Returns false when node does not contain a 'td' element`, () => {
    const child: Element = { type: `element`, tagName: `div`, children: [] };
    const node: Element = { type: `element`, tagName: `tr`, children: [child] };
    const result = defaultPredicateFn(node);

    expect(result).toBeFalsy();
  });

  it(`Returns false when node does not have a child 'td' element whose id attribute matches a prefix`, () => {
    const child: Element = {
      type: `element`,
      tagName: `td`,
      children: [],
      properties: {
        id: `invalid`,
      },
    };
    const node: Element = { type: `element`, tagName: `tr`, children: [child] };
    const result = defaultPredicateFn(node);

    expect(result).toBeFalsy();
  });

  it(`Returns true when node has a child 'td' element whose id attribute matches a prefix`, () => {
    const child: Element = {
      type: `element`,
      tagName: `td`,
      children: [],
      properties: {
        id: `file-filename-LC1`,
      },
    };
    const node: Element = { type: `element`, tagName: `tr`, children: [child] };
    const result = defaultPredicateFn(node);

    expect(result).toBeTruthy();
  });
});

describe(`getGistVisitor`, () => {
  const generateTableRow = (col1Props: Properties = {}, col2Props: Properties = {}): Element => ({
    type: `element`,
    tagName: `tr`,
    children: [
      {
        type: `element`,
        tagName: `td`,
        children: [],
        properties: col1Props,
      },
      {
        type: `element`,
        tagName: `td`,
        children: [],
        properties: col2Props,
      },
    ],
  });
  const generateCodeLineNode = (lineNumber: number): Element => generateTableRow(
    { id: `random` },
    { id: `${getCodeLineIdPrefix(singleFileGist.uri)}${lineNumber}` },
  );
  const executeSut = (uri: GistUri, rootNode: Element) => visit(
    rootNode,
    // getGistVisitorPredicate(uri),
    (node: Element) => isElement(node, `tr`), // using a simpler predicate so more elements get fed to visitor
    getGistVisitor(uri),
    true,
  );

  it(`Returns the Gist HTML with highlighted lines when specified`, () => {
    const lineNumber = 3;
    const uri: GistUri = {
      ...singleFileGist.uri,
      highlights: [lineNumber],
    };
    const node: Element = generateCodeLineNode(lineNumber);
    const parent: Element = {
      type: `element`,
      tagName: `table`,
      children: [node],
    };

    executeSut(uri, parent);

    const highlightedNode = node.children[1] as Element;

    expect(highlightedNode.properties.className).not.toBeUndefined();
    expect(highlightedNode.properties.className).toContain(`highlighted`);
  });

  it(`Returns the Gist HTML with lines removed when specified`, () => {
    const uri: GistUri = {
      ...singleFileGist.uri,
      lines: [2, 3, 5],
    };
    const nodes: Element[] = [...Array(5)].map((_, index) => generateCodeLineNode(index + 1));
    const parent: Element = {
      type: `element`,
      tagName: `table`,
      children: nodes,
    };

    executeSut(uri, parent);

    const lineOne = (parent.children[0] as Element).children[1] as Element;
    const lineFour = (parent.children[1] as Element).children[1] as Element;

    expect(parent.children.length).toBe(2);
    expect(lineOne.properties.id).toBe(`${getCodeLineIdPrefix(uri)}1`);
    expect(lineFour.properties.id).toBe(`${getCodeLineIdPrefix(uri)}4`);
  });

  it(`Does not modify node if id attribute is invalid`, () => {
    const uri: GistUri = {
      ...singleFileGist.uri,
      highlights: [1],
      lines: [1],
    };
    // Must include prefix or test won't get past predicate. Excluding line number to make invalid.
    const node0: Element = generateTableRow(null, { id: getCodeLineIdPrefix(uri) });
    const expected0: Element = JSON.parse(JSON.stringify(node0));
    const parent: Element = {
      type: `element`,
      tagName: `table`,
      children: [node0],
    };

    executeSut(uri, parent);
    expect(node0).toEqual(expected0);

    const node1: Element = generateTableRow();
    const expected1: Element = JSON.parse(JSON.stringify(node1));
    parent.children = [node1];

    executeSut(uri, parent);
    expect(node1).toEqual(expected1);
  });
});

describe(`getGistElementFromHtml`, () => {
  it(`Returns the Gist Element unmodified with only username and gistId provided`, () => {
    const { uri, gist } = singleFileGist;
    const result = getGistElementFromHtml(uri, gist);
    const expected = getGistTreeFromHtml(gist.div);

    expect(result).toEqual(expected);
  });

  it(`Returns the Gist Element unmodified if the file is not provided`, () => {
    const uri: GistUri = {
      ...singleFileGist.uri,
      file: undefined,
      lines: [1],
      highlights: [2],
    };
    const result = getGistElementFromHtml(uri, singleFileGist.gist);
    const expected = getGistTreeFromHtml(singleFileGist.gist.div);

    expect(result).toEqual(expected);
  });

  it(`Returns the Gist Element unmodified if lines and highlights are not provided`, () => {
    const uri: GistUri = {
      ...singleFileGist.uri,
      file: singleFileGist.gist.files[0],
    };
    const result = getGistElementFromHtml(uri, singleFileGist.gist);
    const expected = getGistTreeFromHtml(singleFileGist.gist.div);

    expect(result).toEqual(expected);
  });

  it(`Modifies Gist Element when the file, lines, and highlights are provided`, () => {
    const uri: GistUri = {
      ...singleFileGist.uri,
      lines: [1],
      highlights: [2],
    };
    const result = getGistElementFromHtml(uri, singleFileGist.gist);
    const unexpected = getGistTreeFromHtml(singleFileGist.gist.div);

    expect(result).not.toEqual(unexpected);
  });
});

describe(`getGistElement`, () => {
  it(`Throws an error when Gist URI has incorrect format`, async () => {
    await expect(() => getGistElement(`this is invalid`)).rejects.toThrow(
      `Failed to load Gist, incorrect URI format`,
    );
  });

  it(`Throws an error when the Gist at the specified URI doesn't exist`, async () => {
    axios.get = import.meta.jest.fn().mockResolvedValue({ status: 404 });
    await expect(getGistElement(`gist:randomdude/cj9ht92huryh89rh34h8rh8`)).rejects.toThrow(
      `Gist not found at URL: https://gist.github.com/randomdude/cj9ht92huryh89rh34h8rh8.json`,
    );
  });

  it(`Throws an error when there is a server error while requesting the Gist`, async () => {
    axios.get = import.meta.jest.fn().mockResolvedValue({ status: 500 });
    await expect(getGistElement(`gist:a/b`)).rejects.toThrow(
      `An error occurred while requesting Gist from the server`,
    );
  });

  it.each([
    [300, `x`, `300 - x`],
    [401, `y`, `401 - y`],
    [418, `I'm a little teapot`, `418 - I'm a little teapot`],
  ])(`Throws an error if the Gist request status is unsupported`, async (status, statusText, expected) => {
    axios.get = import.meta.jest.fn().mockResolvedValue({
      status,
      statusText,
    });
    await expect(getGistElement(`gist:x/y`)).rejects.toThrow(`Response not supported: ${expected}`);
  });
});

describe(`elementHasOneValidChild`, () => {
  it(`Returns false when element contains more than one child element`, () => {
    const node: Element = {
      type: `element`,
      tagName: `div`,
      children: [
        {
          type: `element`,
          tagName: `span`,
          children: [],
        },
        {
          type: `element`,
          tagName: `span`,
          children: [],
        },
      ],
    };

    const result = elementHasOneValidChild(node);

    expect(result).toBeFalsy();
  });

  it(`Returns false when element contains only whitespace`, () => {
    const node: Element = {
      type: `element`,
      tagName: `div`,
      children: [
        {
          type: `text`,
          value: `\n\t \r\n \r`,
        },
      ],
    };

    const result = elementHasOneValidChild(node);

    expect(result).toBeFalsy();
  });

  it(`Returns true when element contains one child element`, () => {
    const node: Element = {
      type: `element`,
      tagName: `div`,
      children: [
        {
          type: `element`,
          tagName: `span`,
          children: [],
        },
      ],
    };

    const result = elementHasOneValidChild(node);

    expect(result).toBeTruthy();
  });

  it(`Returns true when element contains one child element and at least one instance of whitespace`, () => {
    const node: Element = {
      type: `element`,
      tagName: `div`,
      children: [
        {
          type: `text`,
          value: `    \n`,
        },
        {
          type: `element`,
          tagName: `span`,
          children: [],
        },
        {
          type: `text`,
          value: `    \n`,
        },
      ],
    };

    const result = elementHasOneValidChild(node);

    expect(result).toBeTruthy();
  });
});

describe(`convertInlineCodeToGist`, () => {
  beforeAll(() => {
    axios.get = import.meta.jest.fn().mockResolvedValue({
      status: 200,
      statusText: `OK`,
      data: singleFileGist.gist,
    });
  });

  const defaultOptions: RehypeGistOptions = {
    replaceParentParagraph: true,
    omitCodeBlocks: true,
  };

  it(`Does not transform elements without text`, async () => {
    const node: Element = {
      type: `element`,
      tagName: `code`,
      children: [],
    };
    const parent: Element = {
      type: `element`,
      tagName: `div`,
      children: [node],
    };
    const expectedNode = JSON.parse(JSON.stringify(node));
    const expectedParent = JSON.parse(JSON.stringify(parent));

    await convertInlineCodeToGist(node, parent, defaultOptions);

    expect(node).toEqual(expectedNode);
    expect(parent).toEqual(expectedParent);
  });

  it(`Does not transform elements that are not 'code' tags`, async () => {
    const node: Element = {
      type: `element`,
      tagName: `span`,
      children: [
        {
          type: `text`,
          value: `gist:a/b`,
        },
      ],
    };
    const parent: Element = {
      type: `element`,
      tagName: `div`,
      children: [node],
    };
    const expectedNode = JSON.parse(JSON.stringify(node));
    const expectedParent = JSON.parse(JSON.stringify(parent));

    await convertInlineCodeToGist(node, parent, defaultOptions);

    expect(node).toEqual(expectedNode);
    expect(parent).toEqual(expectedParent);
  });

  it(`Handles 'code' tag transformations that have 'pre' as direct parent`, async () => {
    const node: Element = {
      type: `element`,
      tagName: `code`,
      children: [
        {
          type: `text`,
          value: `gist:a/b`,
        },
      ],
    };
    const parent: Element = {
      type: `element`,
      tagName: `pre`,
      children: [node],
    };
    const expectedUnmodifiedNode = JSON.parse(JSON.stringify(node));
    const expectedUnmodifiedParent = JSON.parse(JSON.stringify(parent));

    await convertInlineCodeToGist(node, parent, defaultOptions);

    expect(node).toEqual(expectedUnmodifiedNode);
    expect(parent).toEqual(expectedUnmodifiedParent);

    await convertInlineCodeToGist(node, parent, { ...defaultOptions, omitCodeBlocks: false });

    expect(node).not.toEqual(expectedUnmodifiedNode);
    expect(parent).not.toEqual(expectedUnmodifiedParent);
    expect(node.tagName).toBe(`div`);
    expect(node.properties.className).toContain(`gist`);
  });

  it(`Does not replace parent 'p' tag if the 'code' tag has sibling elements`, async () => {
    const node: Element = {
      type: `element`,
      tagName: `code`,
      children: [
        {
          type: `text`,
          value: `gist:a/b`,
        },
      ],
    };
    const siblingNode: Element = {
      type: `element`,
      tagName: `span`,
      children: [],
    };
    const parent: Element = {
      type: `element`,
      tagName: `pre`,
      children: [node, siblingNode],
    };
    const expectedUnmodifiedNode = JSON.parse(JSON.stringify(node));
    const expectedUnmodifiedParent = JSON.parse(JSON.stringify(parent));

    await convertInlineCodeToGist(node, parent, defaultOptions);

    expect(node).toEqual(expectedUnmodifiedNode);
    expect(parent).toEqual(expectedUnmodifiedParent);
  });

  it(`Handles 'code' tag transformations that replace its parent 'p' tag`, async () => {
    const node: Element = {
      type: `element`,
      tagName: `code`,
      children: [
        {
          type: `text`,
          value: `gist:a/b`,
        },
      ],
    };
    const parent: Element = {
      type: `element`,
      tagName: `p`,
      children: [node],
    };
    const unexpectedUnmodifiedParent = JSON.parse(JSON.stringify(parent));

    await convertInlineCodeToGist(node, parent, defaultOptions);

    expect(parent).not.toEqual(unexpectedUnmodifiedParent);
    expect(parent.tagName).toBe(`div`);
    expect(parent.properties.className).toContain(`gist`);
  });

  it(`Handles 'code' tag transformations that do not replace its parent 'p' tag`, async () => {
    const node: Element = {
      type: `element`,
      tagName: `code`,
      children: [
        {
          type: `text`,
          value: `gist:a/b`,
        },
      ],
    };
    const parent: Element = {
      type: `element`,
      tagName: `p`,
      children: [node],
    };
    const unexpectedUnmodifiedNode = JSON.parse(JSON.stringify(node));
    const unexpectedUnmodifiedParent = JSON.parse(JSON.stringify(parent));

    await convertInlineCodeToGist(node, parent, {
      ...defaultOptions,
      replaceParentParagraph: false,
    });

    expect(node).not.toEqual(unexpectedUnmodifiedNode);
    expect(parent).not.toEqual(unexpectedUnmodifiedParent);
    expect(node.tagName).toBe(`div`);
    expect(node.properties.className).toContain(`gist`);
    expect(parent.tagName).toBe(`p`);
  });

  it(`Transforms 'code' element to Gist`, async () => {
    const node: Element = {
      type: `element`,
      tagName: `code`,
      children: [
        {
          type: `text`,
          value: `gist:a/b`,
        },
      ],
    };
    const parent: Element = {
      type: `element`,
      tagName: `div`,
      children: [node],
    };
    const unexpectedUnmodifiedNode = JSON.parse(JSON.stringify(node));
    const unexpectedUnmodifiedParent = JSON.parse(JSON.stringify(parent));

    await convertInlineCodeToGist(node, parent, defaultOptions);

    expect(node).not.toEqual(unexpectedUnmodifiedNode);
    expect(parent).not.toEqual(unexpectedUnmodifiedParent);
    expect(node.tagName).toBe(`div`);
    expect(node.properties.className).toContain(`gist`);
  });
});

describe(`isValidGist`, () => {
  it(`Is valid only if element is a 'code' tag`, () => {
    const validNode: Element = {
      type: `element`,
      tagName: `code`,
      children: [{
        type: `text`,
        value: `gist:a/b`,
      }],
    };
    const invalidNode: Element = {
      type: `element`,
      tagName: `span`,
      children: [{
        type: `text`,
        value: `gist:a/b`,
      }],
    };

    expect(isValidGist(validNode)).toBeTruthy();
    expect(isValidGist(invalidNode)).toBeFalsy();
  });

  it(`Is invalid if element has more than one child`, () => {
    const invalidNode: Element = {
      type: `element`,
      tagName: `code`,
      children: [{
        type: `text`,
        value: `gist:a/b`,
      },
      {
        type: `text`,
        value: `gist:a/b`,
      }],
    };

    expect(isValidGist(invalidNode)).toBeFalsy();
  });

  it(`Is invalid if child of element is not of type 'text'`, () => {
    const invalidNode: Element = {
      type: `element`,
      tagName: `code`,
      children: [{
        type: `element`,
        tagName: `span`,
        children: [
          {
            type: `text`,
            value: `gist:a/b`,
          },
        ],
      }],
    };

    expect(isValidGist(invalidNode)).toBeFalsy();
  });

  it(`Is valid if element text begins with 'gist:' protocol`, () => {
    const invalidNode: Element = {
      type: `element`,
      tagName: `code`,
      children: [{
        type: `text`,
        value: `gisp:a/b`,
      }],
    };

    expect(isValidGist(invalidNode)).toBeFalsy();
  });
});


test/index.test.ts
describe(`convertInlineCodeToGist`, () => {
  it(`Does not transform elements that are not 'code' tags`, () => {

  });

  it(`Does not transform 'code' tags that have 'pre' as direct parent`, () => {

  });

  it(`Does not replace parent 'p' tag if the 'code' tag has sibling elements`, () => {

  });

  it(`Throws an error if root Gist 'div' element has siblings`, () => {

  });

  it(`Replaces parent 'p' tag when transforming`, () => {

  });

  it(`Transforms 'code' element to Gist`, () => {

  });
});

describe(`isValidGist`, () => {
  it(`Is valid if element is a 'code' tag`, () => {

  });

  it(`Is valid if element has a single child`, () => {

  });

  it(`Is valid if child of element is of type 'text'`, () => {

  });

  it(`Is valid if element text begins with 'gist:' protocol`, () => {

  });
});


src/util.ts
import qs from "qs";
import axios from "axios";
import parseRange from "parse-numeric-range";
import { fromHtml } from "hast-util-from-html";
import type {
  Element,
  Parent,
  Root,
  Text,
} from "hast";
import { isElement, TestFunctionAnything } from "hast-util-is-element";
import { visit, Visitor } from "unist-util-visit";
import { classnames as cx } from "hast-util-classnames";

export type GistUri = {
  username: string;
  gistId: string;
  file?: string;
  lines: number[];
  highlights: number[];
};

export type Gist = {
  description: string;
  public: boolean;
  created_at: string;
  files: string[];
  owner: string;
  div: string;
  stylesheet: string;
};

export type RehypeGistOptions = {
  replaceParentParagraph?: boolean;
  omitCodeBlocks?: boolean;
  classNames?: string | string[];
};

const baseGistUrl = `https://gist.github.com`;
const gistProtocol = `gist:`;

export function parseGistUri(gistUri: string): GistUri | undefined {
  if (!gistUri.startsWith(gistProtocol)) return undefined;

  const uriQuery = gistUri.replace(gistProtocol, ``).split(`?`);

  if (uriQuery.length > 2) return undefined;

  const [uriPath, queryString] = uriQuery;

  const uriSegments = uriPath.split(`/`);

  if (uriSegments.length === 2 && !queryString) {
    return {
      username: uriSegments[0],
      gistId: uriSegments[1],
      lines: [],
      highlights: [],
    } as GistUri;
  }

  if (uriSegments.length !== 2) return undefined;

  const [username, gistId] = uriSegments;
  const gistQuery = qs.parse(queryString, { comma: false });
  const highlightsQuery = gistQuery.highlights?.toString();
  const highlights = highlightsQuery ? parseRange(highlightsQuery) : [];
  const linesQuery = gistQuery.lines?.toString();
  const lines = linesQuery ? parseRange(linesQuery) : [];

  return {
    username,
    gistId,
    file: gistQuery.file?.toString(),
    lines,
    highlights,
  };
}

export function convertFilenameToAttributeFormat(filename: string): string {
  return filename.replace(/^\./, ``)
    .replace(/[^a-zA-Z0-9_]+/g, `-`)
    .toLowerCase();
}

export function getCodeLineIdPrefix(uri: GistUri): string {
  return `file-${convertFilenameToAttributeFormat(uri.file)}-LC`;
}

export function getGistTreeFromHtml(gistHtml: string): Element {
  const rootFragment: Root = fromHtml(gistHtml, { fragment: true });
  const rootElements: Element[] = rootFragment.children
    .filter((child) => isElement(child))
    .map((child) => child as Element);

  if (rootElements.length !== 1) throw new Error(`Gist doesn't exist or has invalid content`);

  return rootElements[0];
}

export function getGistVisitorPredicate(uri: GistUri): TestFunctionAnything {
  return (node: Element) => {
    if (!isElement(node, `tr`)) return false;

    const elementIdPrefix = getCodeLineIdPrefix(uri);
    const codeLine = node.children.find(
      (e: Element) => e.tagName === `td` && String(e.properties?.id).includes(elementIdPrefix),
    );

    return codeLine !== undefined;
  };
}

export function getGistVisitor(uri: GistUri): Visitor {
  return (node: Element, index: number, parent: Parent) => {
    const elementIdPrefix = getCodeLineIdPrefix(uri);
    const codeLine = node.children.find(
      (e: Element) => String(e.properties?.id).includes(elementIdPrefix),
    );

    if (codeLine === undefined) return;

    const lineNumber = parseInt(
      String((codeLine as Element).properties.id).replace(elementIdPrefix, ``),
      10,
    );

    if (Number.isNaN(lineNumber)) return;

    if (uri.highlights.includes(lineNumber)) {
      Object.assign(codeLine, cx(codeLine, `highlighted`));
    }

    if (uri.lines.includes(lineNumber)) {
      parent.children.splice(index, 1);
    }
  };
}

export function getGistElementFromHtml(uri: GistUri, data: Gist): Element {
  const gistHtml = data.div;
  const hasHighlights = uri.highlights.length > 0;
  const hasLines = uri.lines.length > 0;
  const canProcessLinesOrHighlights = (hasLines || hasHighlights) && uri.file;
  const gistElement = getGistTreeFromHtml(gistHtml);

  if (!canProcessLinesOrHighlights) return gistElement;

  visit(
    gistElement,
    getGistVisitorPredicate(uri),
    getGistVisitor(uri),
    true,
  );

  return gistElement;
}

export default async function getGistElement(gistUri: string): Promise<Element> {
  const uri = parseGistUri(gistUri);

  if (!uri) throw new Error(`Failed to load Gist, incorrect URI format`);

  const gistUrl = `${baseGistUrl}/${uri.username}/${uri.gistId}.json${uri.file ? `?file=${uri.file}` : ``}`;
  const response = await axios.get(gistUrl);

  if (response.status === 404) throw new Error(`Gist not found at URL: ${gistUrl}`);
  if (response.status === 500) throw new Error(`An error occurred while requesting Gist from the server`);
  if (response.status !== 200) throw new Error(`Response not supported: ${response.status} - ${response.statusText}`);

  return getGistElementFromHtml(uri, response.data);
}

export function isValidGist(node: Element): boolean {
  return isElement(node, `code`)
    && node.children.length === 1
    && node.children[0].type === `text`
    && node.children[0].value.startsWith(gistProtocol);
}

export function elementHasOneValidChild(node: Element): boolean {
  let elementCount = 0;

  node.children.forEach((child) => {
    if (isElement(child)
      || (child.type === `text`
        && (child as Text).value.search(/^\s+$/) === -1)
    ) elementCount += 1;
  });

  return elementCount === 1;
}

export async function convertInlineCodeToGist(
  node: Element,
  parent: Parent,
  options: RehypeGistOptions,
): Promise<void> {
  if (!isValidGist(node)) return;
  if (options.omitCodeBlocks && isElement(parent, `pre`)) return;
  // Can't replace p tag if it contains more than just our target gist element
  if (options.replaceParentParagraph
    && isElement(parent, `p`)
    && !elementHasOneValidChild(parent)) return;

  const gistUri = (node.children[0] as Text).value;
  const gistElement = await getGistElement(gistUri);
  const newNode: Element = cx(gistElement, options.classNames);

  if (options.replaceParentParagraph && isElement(parent) && isElement(parent, `p`)) {
    Object.assign(parent, newNode);
  } else {
    Object.assign(node, newNode);
  }
}

export function getGistTransformations(tree: Root, options: RehypeGistOptions): Promise<void>[] {
  const transformations: Promise<void>[] = [];
  visit(
    tree,
    isValidGist,
    (node: Element, _, parent: Parent) => {
      transformations.push(
        convertInlineCodeToGist(node, parent, options),
      );
    },
  );

  return transformations;
}
